### DISCUSSION TITLE
Checking if local repository is sane with python bindings

### DISCUSSION THREAD

**Discussion by mattiaverga** [rpm-software-management/dnf5#1946] [Q&A] [answered]
I'm [trying](https://github.com/fedora-infra/bodhi/pull/5822) to rewrite Fedora Bodhi's repository sanity check using python-libdnf5 bindings.
The scope is just to try loading a local repository that was previously created to check if anything is broken. Currently Bodhi does that by firing a `list --available` command through subprocess, but that [seems to not work anymore with dnf5](https://github.com/fedora-infra/bodhi/issues/5820) (possibly because of privileges problem running the command from an unprivileged user).

Python docs and tutorial are not really clear to me, so I've wrote this:
```
def load_repo_libdnf5(tempdir, myurl):
    """
    Use libdnf5 python bindings to try to load a repository.
    Args:
        tempdir (str): Temporary directory for libdnf cache.
        myurl (str): A path to a repodata directory.
    Raises:
        Exception: If the repodata is not valid or does not exist.
    """
    base = libdnf5.base.Base()
    base_config = base.get_config()
    base_config.plugins = False
    base_config.cachedir = tempdir
    base.setup()
    repo_sack = base.get_repo_sack()
    repo = repo_sack.create_repo("testrepo")
    repo.get_config().baseurl = myurl
    repo_sack.load_repos(libdnf5.repo.Repo.Type_AVAILABLE)
    return True
```
The question is: using `load_repos` is enough to be sure the created repository is sane? And: I don't see how `base_config` is applied when running `base.setup()`, is it?

I also thought it would be nice to check if the loaded repository is actually `testrepo`, so I've added:
```
query = libdnf5.repo.RepoQuery(base)
query.filter_enabled(True)
for r in query:
    print(r.get_name())
```
but that returns an empty string...

**Comment by kontura** [comment, marked as answer]
> The question is: using `load_repos` is enough to be sure the created repository is sane? 

It will load the repositories so if it doesn't throw the repos are valid. However it doesn't tell you about the contents, whether there are any groups, advisories, packages... or whether package dependencies can be satisfied and stuff like that.

> And: I don't see how `base_config` is applied when running `base.setup()`, is it?

Yes, for example you can checkout: https://github.com/rpm-software-management/dnf5/blob/main/libdnf5/base/base.cpp#L181-L212 where it uses the `config` right at the beginning of `setup`.

> but that returns an empty string...

It is because the repository doesn't have a `name` set. To set it you need to do something like `repo.get_config().name = "Test Repo Name"`.
It does have an `id` though so maybe you are looking for `print(r.get_id())`.

It reflects the repo file structure:
```
[testrepo]
name=Test Repo Name
baseurl=myurl
```

**Reply by mattiaverga** [reply]
> It will load the repositories so if it doesn't throw the repos are valid. However it doesn't tell you about the contents, whether there are any groups, advisories, packages... or whether package dependencies can be satisfied and stuff like that.

Yes, we just need to be sure the repo data is loadable.
